dttools/src/catalog_query.c

```diff
@@ -98,53 +120,43 @@ struct catalog_query *catalog_query_create(const char *hosts, struct jx *filter_
 		}
 	} while (next_host);
 
-	sorted_hosts = list_splice(previously_up, previously_down);
+	return list_splice(previously_up, previously_down);
+}
+
+struct catalog_query *catalog_query_create(const char *hosts, struct jx *filter_expr, time_t stoptime)
+{
+	struct catalog_query *q = NULL;
+	char *n;
+	struct catalog_host *h;
+	struct list *sorted_hosts = catalog_query_sort_hostlist(hosts);
 
 	list_first_item(sorted_hosts);
 	while((h = list_next_item(sorted_hosts))) {
-		struct link *link = http_query(h->url, "GET", stoptime);
-
-		if(!link) {
-			goto FAILURE;
-		}
-
-		struct jx *j = jx_parse_link(link,stoptime);
-
-		link_close(link);
-
-		if(!j) {
-			debug(D_DEBUG,"query result failed to parse as JSON");
-			goto FAILURE;
-		}
-
-		if(!jx_istype(j,JX_ARRAY)) {
-			debug(D_DEBUG,"query result is not a JSON array");
-			jx_delete(j);
-			goto FAILURE;
-		}
-
-		q = xxmalloc(sizeof(*q));
-		q->data = j;
-		q->current = j->u.items;
-		q->filter_expr = filter_expr;
-
-		if(h->down) {
-			debug(D_DEBUG,"catalog server at %s is back up", h->host);
-			set_first_element(down_hosts);
-			while((n = set_next_element(down_hosts))) {
-				if(!strcmp(n, h->host)) {
-					free(n);
-					set_remove(down_hosts, n);
-					break;
+		struct jx *j = catalog_query_send_query(h->url, stoptime);
```

Comment:
Huh, I had assumed `stoptime` worked as a timeout. Looking at `link.h`, I see that is not the case.

When a stop time is specified, should we divide up the total time among the servers, or find the relative time and let each request take the same? If we divide the time up, a large number of servers could push the time for each request too low for any of them to succeed. Since the link API uses absolute times, if we give each request the relative time specified, callers might be surprised that the query runs past the absolute stoptime.